subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiAccessTest.java

```diff
@@ -66,13 +70,20 @@ public class PublicApiAccessTest {
                 .or(type(Closure.class))
                 .as("Groovy classes")
             );
+    private static final DescribedPredicate<JavaClass> public_api_tasks_or_plugins =
+            gradlePublicApi().and(assignableTo(Task.class).or(assignableTo(Plugin.class)));
 
     @ArchTest
     public static final ArchRule public_api_methods_do_not_reference_internal_types_as_parameters = freeze(methods()
         .that(are(public_api_methods))
         .should(haveOnlyArgumentsOrReturnTypesThatAre(allowed_types_for_public_api))
     );
 
+    @ArchTest
+    public static final ArchRule public_api_tasks_and_plugins_are_abstract = freeze(classes()
```

Comment:
If there aren't any exceptions to the rule, I wouldn't add `freeze` around it. So for the moment, you wouldn't be able to accept exceptions, though I think that is fine.